src/ucx_rma_plugin.c

```diff
@@ -26,1239 +23,1277 @@
-    req_param.memory_type  =  mh->mem_type;
+  if (rtr->avail == rtr->count) {
+    assert(atp->reqs == 0);
+    assert(atp->inflight == 0);
+    atp->id_start = rtr->id_start;
+    atp->count    = 0;
+    memset(atp->sizes, 0, sizeof(atp->sizes));
+    atp->id = rtr->id_start;
   }
-  
-  st  = ucp_put_nbx(comm->ep, data, size, slot->addr,
-                    comm->rkeys[*rkey_index].rkey,
-                    &req_param);
-
-  if (UCS_PTR_IS_ERR(st)) {
-    WARN("NET/UCX_RMA: isend pub_nb failed");
-    return ncclInternalError;
-  } else if (st  == NULL) {
-    req->done |= NCCL_UCX_RMA_REQUEST_PUT_DONE;
-  }
-
-  ucp_worker_fence(comm->super.worker);
-  req->am_msg = (((uint64_t)slot->req_id) << 32) | ((uint64_t)size);
-  req->st = ucp_am_send_nb(comm->ep, comm->rem_am_id, &req->am_msg, 8,
-                           ucp_dt_make_contig(1), nccl_ucx_rma_am_isend_cb, 0);
 
-  if (req->st == NULL) {
-    req->done |= NCCL_UCX_RMA_REQUEST_AM_DONE;
-  } else if (UCS_PTR_IS_PTR(req->st)) {
-    nccl_ucx_am_request_t *am_req = (nccl_ucx_am_request_t*)req->st;
-    am_req->req = req;
-  } else {
-    WARN("NET/UCX_RMA: isend am_send_nb failed");
-  }
+  req->comm      = comm;
+  req->type      = NCCL_UCP_TYPE_ISEND;
+  req->rtr_id    = rtr->id_start;
+  req->inflight  = (UCS_PTR_STATUS(status_ptr) == UCS_INPROGRESS);
+  atp->inflight += req->inflight;
+  atp->sizes[i]  = size;
+  atp->count++;
+  atp->reqs++;
 
-  req->seq = slot->seq;
-  slot->ready = 0;
-  slot->addr  = 0ULL;
-  slot->size  = 0;
-  slot->seq   = 0;
-  comm->fifo_head++;
+  rtr->avail--;
+  rtr->chunk[i].tag = INT_MAX;
 
-  req->worker = comm->super.worker;
-  req->type   = UCX_RMA_REQ_TYPE_SEND;
+  comm->req_id++;
+  comm->total++;
   *request = req;
   return ncclSuccess;
 }
 
-static void nccl_ucx_rma_dummy_am_cb(void *request, ucs_status_t status)
-{
-  return;
-}
+static ncclResult_t nccl_ucx_rma_isend(void *send_comm, void *data, int size,
+                                       int tag, void *mhandle, void **request) {
+  ncclResult_t result   = ncclSuccess;
+  nccl_ucp_comm_t *comm = send_comm;
+  volatile nccl_ucp_rtr_t *rtr;
+  unsigned short id;
+  unsigned i;
+
+  *request = NULL;
 
-ncclResult_t nccl_ucx_rma_post_fifo(nccl_ucx_rma_recv_comm_t *comm,
-                                    ucx_rma_mhandle_t *mh,
-                                    uint64_t addr, int size, int req_id)
-{
-  ucx_rma_send_fifo_t    *local_elem;
-  nccl_ucx_rma_request_t *req;
-  uint64_t               remote_addr;
-  ucs_status_t           st;
-
-  if (!mh->rkey_buf.send) {
-    req = &(comm->super.reqs[req_id]);
-    req->st = ucp_am_send_nb(comm->ep, comm->rem_am_id, &mh->rkey_buf,
-                             sizeof(nccl_ucx_rma_rkey_buf_t), ucp_dt_make_contig(1),
-                             nccl_ucx_rma_dummy_am_cb, 0);
-    if (UCS_PTR_IS_ERR(req->st)) {
-      WARN("NET/UCX_RMA: am_send_nb failed");
-      return ncclInternalError;
+  assert(tag != INT_MAX);
```

Comment:
why?

Reply:
we use this value to mark entries as consumed later